src/runtime/future/vanilla_future.c

```diff
@@ -0,0 +1,168 @@
+  bool r;
+  r = __atomic_load_n(&fut->fulfilled, __ATOMIC_SEQ_CST);
+  return r;
+}
+
+static inline void vanilla_future_gc_trace_value(pony_ctx_t *ctx, vanilla_future_t *fut)
+{
+  assert(fut);
+  if (fut->type == ENCORE_ACTIVE) {
+    encore_trace_actor(ctx, fut->value.p);
+  } else if (fut->type != ENCORE_PRIMITIVE) {
+    encore_trace_object(ctx, fut->value.p, fut->type->trace);
+  }
+}
+
+static inline void vanilla_future_gc_send_value(pony_ctx_t *ctx, vanilla_future_t *fut)
+{
+  pony_gc_send(ctx);
+  vanilla_future_gc_trace_value(ctx, fut);
+  pony_send_done(ctx);
+}
+
+static inline void vanilla_future_gc_recv_value(pony_ctx_t *ctx, vanilla_future_t *fut)
+{
+  pony_gc_recv(ctx);
+  vanilla_future_gc_trace_value(ctx, fut);
+  ponyint_gc_handlestack(ctx);
+}
+
+vanilla_future_t *vanilla_future_mk(pony_ctx_t **ctx, pony_type_t *type)
+{
+  pony_ctx_t *cctx = *ctx;
+  assert(cctx->current);
+  vanilla_future_t *fut = pony_alloc_final(cctx, sizeof(future_t),
+          (void *)&vanilla_future_finalizer);
+  *fut = (vanilla_future_t) { .type = type, .blocking = false };
+  ENC_DTRACE3(FUTURE_CREATE, (uintptr_t) ctx, (uintptr_t) fut, (uintptr_t) type);
+  return fut;
+} 
+
+encore_arg_t vanilla_future_get_actor(pony_ctx_t **ctx, vanilla_future_t *fut)
+{
+  if (!__atomic_load_n(&(fut->fulfilled), __ATOMIC_SEQ_CST)) {
+    ENC_DTRACE2(FUTURE_BLOCK, (uintptr_t) *ctx, (uintptr_t) fut);
+    vanilla_future_block_actor(ctx, fut);
+    ENC_DTRACE2(FUTURE_UNBLOCK, (uintptr_t) *ctx, (uintptr_t) fut);
+  }
+  ENC_DTRACE2(FUTURE_GET, (uintptr_t) *ctx, (uintptr_t) fut);
+  return fut->value;
+}
+
+static inline void vanilla_future_block_actor(pony_ctx_t **ctx, vanilla_future_t *fut)
+{
+  perr("future_block_actor");
+  if (__atomic_load_n(&(fut->fulfilled), __ATOMIC_SEQ_CST)) {
+    return;
+  }
+  pony_ctx_t* cctx = *ctx;
+  pony_actor_t *a = cctx->current;
+  pony_unschedule(cctx, a);
+  encore_actor_t *actor = (encore_actor_t*) a;
+  future_tnode_info_t * pony_node = POOL_ALLOC(future_tnode_info_t);
+  *pony_node = (future_tnode_info_t) { .fclass = VANILLA_FUTURE, .fut = (void*)fut };
+  actor_block(ctx, actor, (void *)pony_node);          
+}
+
+void vanilla_future_fulfil(pony_ctx_t **ctx, vanilla_future_t *fut, encore_arg_t value) 
+{
+  assert(fut->fulfilled == false);
+  ENC_DTRACE2(FUTURE_FULFIL_START, (uintptr_t) *ctx, (uintptr_t) fut);
+  fut->value = value;
+  __atomic_store_n(&(fut->fulfilled), true, __ATOMIC_SEQ_CST);
+  vanilla_future_gc_send_value(*ctx, fut);
+  vanilla_future_discharge(ctx, fut);
+  ENC_DTRACE2(FUTURE_FULFIL_END, (uintptr_t) *ctx, (uintptr_t) fut);
+}
+
+static inline void vanilla_future_discharge(pony_ctx_t **ctx, vanilla_future_t *fut) {
+  if (__sync_bool_compare_and_swap(&(fut->blocking), true, false)) { 
+     perr("Unblocking");
+     actor_set_resume((encore_actor_t *)fut->actor);
+     pony_schedule(*ctx, fut->actor);
+  }   
+}
+
+void handle_future(encore_actor_t *actor, pony_ctx_t *futctx, void * pony_node)
```

Comment:
this function seems to be used for all futures. Is this correct?
If so, shouldn't it be placed in a more general future file, instead of in the vanilla future?
(I may be missing the point where this is called and if this is always called)